pkg/proxy/metrics/metrics.go

```diff
@@ -237,26 +238,42 @@ var (
 var registerMetricsOnce sync.Once
 
 // RegisterMetrics registers kube-proxy metrics.
-func RegisterMetrics() {
+func RegisterMetrics(mode kubeproxyconfig.ProxyMode) {
 	registerMetricsOnce.Do(func() {
+		// Core kube-proxy metrics for all backends
 		legacyregistry.MustRegister(SyncProxyRulesLatency)
-		legacyregistry.MustRegister(SyncFullProxyRulesLatency)
-		legacyregistry.MustRegister(SyncPartialProxyRulesLatency)
+		legacyregistry.MustRegister(SyncProxyRulesLastQueuedTimestamp)
 		legacyregistry.MustRegister(SyncProxyRulesLastTimestamp)
-		legacyregistry.MustRegister(NetworkProgrammingLatency)
 		legacyregistry.MustRegister(EndpointChangesPending)
 		legacyregistry.MustRegister(EndpointChangesTotal)
 		legacyregistry.MustRegister(ServiceChangesPending)
 		legacyregistry.MustRegister(ServiceChangesTotal)
-		legacyregistry.MustRegister(IptablesRulesTotal)
-		legacyregistry.MustRegister(IptablesRulesLastSync)
-		legacyregistry.MustRegister(IptablesRestoreFailuresTotal)
-		legacyregistry.MustRegister(IptablesPartialRestoreFailuresTotal)
-		legacyregistry.MustRegister(SyncProxyRulesLastQueuedTimestamp)
-		legacyregistry.MustRegister(SyncProxyRulesNoLocalEndpointsTotal)
 		legacyregistry.MustRegister(ProxyHealthzTotal)
 		legacyregistry.MustRegister(ProxyLivezTotal)
 
+		// FIXME: winkernel does not implement these
+		legacyregistry.MustRegister(NetworkProgrammingLatency)
+		legacyregistry.MustRegister(SyncProxyRulesNoLocalEndpointsTotal)
+
+		switch mode {
+		case kubeproxyconfig.ProxyModeIPTables:
+			legacyregistry.MustRegister(SyncFullProxyRulesLatency)
+			legacyregistry.MustRegister(SyncPartialProxyRulesLatency)
+			legacyregistry.MustRegister(IptablesRestoreFailuresTotal)
+			legacyregistry.MustRegister(IptablesPartialRestoreFailuresTotal)
+			legacyregistry.MustRegister(IptablesRulesTotal)
+			legacyregistry.MustRegister(IptablesRulesLastSync)
+
+		case kubeproxyconfig.ProxyModeIPVS:
+			legacyregistry.MustRegister(IptablesRestoreFailuresTotal)
+
+		case kubeproxyconfig.ProxyModeNFTables:
+			// FIXME: should not use the iptables-specific metric
+			legacyregistry.MustRegister(IptablesRestoreFailuresTotal)
+
+		case kubeproxyconfig.ProxyModeKernelspace:
+			// currently no winkernel-specific metrics
+		}
```

Comment:
@dgrisonnet since you helped with kube-proxy metrics on another PR...

does this make sense? It seemed wrong to me to be registering metrics in modes where they don't get used (eg, registering the "iptables-restore failure" metric in windows or nftables mode), but maybe there's a rule that says that we should always register all the same metrics, even if some of them will always be 0?

Reply:
There is no such rule, registering conditionally like you did sounds like a cleaner approach to me